### PR TITLE
kernel: Improvements to process cleanup.

### PR DESCRIPTION
**Pull request by Steveice10**
* Reorganize ExitProcess code to be a bit cleaner and reusable for TerminateProcess.
* Fixes to termination of threads on process exit:
  * Terminate threads on all cores.
  * Make sure terminated threads are removed from the thread_list so they can be released.
  * Allow terminating non-waiting threads. This is probably not exactly how the real kernel does it, so I've left a TODO item, but it works for the most part. Better to work at least some of the time than always fail the assert.
* Clean up lingering memory allocations on process destruction.
  * Resolves cases where opening a game after closing another would attempt bogus linear heap allocations, due to incorrect resource limit memory commit values and a lack of free memory in the application region.
* Implement home menu path for APT:DoApplicationJump, which works now with the above changes.

With these changes a number of games (e.g. Super Mario 3D Land, Pokemon Mystery Dungeon: Gates to Infinity) can be quit and re-launched from the home menu instead of crashing.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6680)

### CODE REVIEW

File: src/core/hle/kernel/thread.cpp [hunk 1]
```diff
@@ -199,11 +199,33 @@ void ThreadManager::WaitCurrentThread_Sleep() {
...
     thread->Stop();
-    thread_list.erase(std::remove_if(thread_list.begin(), thread_list.end(),
-                                     [thread](const auto& p) { return p.get() == thread; }),
-                      thread_list.end());
+    std::erase(thread_list, thread);
+}
+
+void ThreadManager::TerminateProcessThreads(std::shared_ptr<Process> process) {
+    auto iter = thread_list.begin();
+    while (iter != thread_list.end()) {
```
Comment (GPUCode) [inline src/core/hle/kernel/thread.cpp:213]: This loop can probably be implemeted using std::erase_if

**Review by GPUCode — APPROVED** [review]

File: src/core/hle/kernel/thread.cpp (hunk 1 above)
Comment (Steveice10) [reply src/core/hle/kernel/thread.cpp:213]: We also need to call thread->Stop() within the loop

Comment (GPUCode) [reply src/core/hle/kernel/thread.cpp:213]: Was thinking something like this
```
std::erase_if(thread_list, [&](auto&& thread) {
    if (thread == current_thread || thread->owner_process.lock() != process) {
        return false;
    }
    thread->Stop();
    return true;
});
```

**Comment by Steveice10** [comment]
Made some adjustments to the process memory cleanup so we can properly track what memory the process owns instead of going through all mapped memory. Makes things a bit cleaner and more accurate, and resolves a crash when stopping emulation from the home menu.

**Review by FearlessTobi — APPROVED** [review]
Looks good, let's leave this for a bit in Canary please.

**Pull request merged**